### PR TITLE
Improve List.keysort/3 in terms of code simplication and guard checking

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

File: lib/elixir/lib/list.ex [hunk 1]
```diff
@@ -491,33 +497,33 @@ defmodule List do
...
-  def keysort(list, position, sorter) when is_list(list) and is_integer(position) do
+  def keysort_guarded(list, position, sorter) do
     :lists.sort(keysort_fun(sorter, position + 1), list)
   end
 
   defp keysort_fun(sorter, position) when is_function(sorter, 2),
     do: &sorter.(:erlang.element(position, &1), :erlang.element(position, &2))
 
-  defp keysort_fun(:asc, position),
-    do: &(:erlang.element(position, &1) <= :erlang.element(position, &2))
```
Comment (eksperimental) [inline lib/elixir/lib/list.ex:507]: This clause is deleted as it is never reached.

**Comment by josevalim** [comment]
Thank you. The title says optimize, do you have any benchmarks?

**Comment by eksperimental** [comment]
> Thank you. The title says optimize, do you have any benchmarks?

It is avoiding unnecessary guard checks

**Comment by eksperimental** [comment]
Well, It is also introducing some new guard checks, so it may not necessarily be an optimization

**Comment by josevalim** [comment]
Oh, the guard checks are not going to slow down and I would expect the compiler to be able to optimize and remove most redundancy. So we should always target code clarity and I think the current version is clearer. A PR to remove the dead code though is very welcome. :)

**Comment by eksperimental** [comment]
Ok. So what about enforcing the guard checks in the public function?

**Comment by josevalim** [comment]
The only benefit of enforcing at the front would be to get function clause errors with diff.  But I Don’t think showing the guard as written in this request would be useful. :(

**Comment by eksperimental** [comment]
True. it's really bad.

**Pull request closed without merging**